bson/src/test/unit/org/bson/codecs/pojo/RecordCustomTest.java

```diff
@@ -0,0 +1,279 @@
+        decodingShouldFail(getRecordCodec(SimpleRecord.class), "{'_t': 'SimpleRecord', 'stringField': 123}");
+    }
+
+    @Test(expected = CodecConfigurationException.class)
+    public void testInvalidTypeForPrimitiveField() {
+        decodingShouldFail(getRecordCodec(PrimitivesRecord.class), "{ '_t': 'PrimitivesRecord', 'myBoolean': null}");
+    }
+
+    @Test(expected = CodecConfigurationException.class)
+    public void testInvalidTypeForRecordField() {
+        decodingShouldFail(getRecordCodec(SimpleNestedRecord.class), "{ '_t': 'SimpleNestedRecord', 'simple': 123}");
+    }
+
+    @Test(expected = CodecConfigurationException.class)
+    public void testInvalidDiscriminatorInNestedRecord() {
+        decodingShouldFail(getRecordCodec(SimpleNestedRecord.class), "{ '_t': 'SimpleNestedRecord',"
+                + "'simple': {'_t': 'FakeRecord', 'integerField': 42, 'stringField': 'myString'}}");
+    }
+
+    @Test(expected = CodecConfigurationException.class)
+    public void testCannotEncodeUnspecializedClasses() {
+        ClassModelBuilder<GenericTreeRecord> builder = ClassModel.builder(GenericTreeRecord.class);
+        CodecRegistry registry = fromProviders(getRecordProviderFromClassModel(builder));
+        encode(registry.get(GenericTreeRecord.class), RecordRoundTripTest.getGenericTreeRecord(), false);
+    }
+
+    @Test(expected = CodecConfigurationException.class)
+    public void testCannotDecodeUnspecializedClasses() {
+        decodingShouldFail(getRecordCodec(GenericTreeRecord.class),
+                "{'field1': 'top', 'field2': 1, "
+                        + "'left': {'field1': 'left', 'field2': 2, 'left': {'field1': 'left', 'field2': 3}}, "
+                        + "'right': {'field1': 'right', 'field2': 4, 'left': {'field1': 'left', 'field2': 5}}}");
+    }
+
+    @Test(expected = CodecConfigurationException.class)
+    public void testInvalidBsonRepresentationStringDecoding() {
+        decodingShouldFail(getRecordCodec(BsonRepresentationUnsupportedStringRecord.class), "{'id': 'hello', s: 3}");
+    }
+
+    @Test(expected = CodecConfigurationException.class)
+    public void testInvalidBsonRepresentationStringEncoding() {
+        encodesTo(getRecordCodecRegistry(BsonRepresentationUnsupportedStringRecord.class),
+                new BsonRepresentationUnsupportedStringRecord("1", "1"), "");
+    }
+
+    @Test(expected = CodecConfigurationException.class)
+    public void testInvalidBsonRepresentationIntDecoding() {
+        decodingShouldFail(getRecordCodec(BsonRepresentationUnsupportedIntRecord.class), "{'id': 'hello', s: '3'}");
+    }
+
+    @Test(expected = CodecConfigurationException.class)
+    public void testStringIdIsNotObjectId() {
+        encodesTo(getRecordCodecRegistry(BsonRepresentationUnsupportedIntRecord.class), new BsonRepresentationRecord("notanobjectid", 1), null);
+    }
+
+    private static <T> RecordCodecProvider getRecordProviderFromClassModel(ClassModelBuilder<T> classModelBuilder) {
```

Comment:
These helpers are in this class as opposed to PojoTestCase because I wanted to keep all record related code out of PojoTestCase in case the record test code needs to go in a separate module. The way I organized it is helpers specific to RecordCustomTest go in RecordCustomTest, helpers specific to RecordRoundTripTest go in RecordRoundTripTest, and tests shared by both of them go in RecordRoundTrip and are package private.